File: mg01/game/tech/Assets/Resources/LuaScript/src/orgu/res/model/GameObjCache.lua
--gameobject缓存
--@author jr.zeng
--2017年11月9日 下午8:15:39
local modname = "GameObjCache"
--==================global reference=======================

local csAssetCache = mg.org.AssetCache.me
local csGameObjCache = mg.org.GameObjCache.me

local type = type
local safe_call = safe_call

--===================namespace========================
local ns = "org"
local using = {--[[namespace1,...]]}
local _ENV = namespace(ns, using, modname)
--===================module property========================
local super = CCModule	--父类
GameObjCache = class(modname, super, _ENV)
local GameObjCache = GameObjCache
local self = GameObjCache


local Refer = Refer
local assetCache = AssetCache


--===================module content========================

function GameObjCache:__ctor()

    self.m_notifier = new(Subject)  --独立派发器
    
end

function GameObjCache:__setup()

    self:ctor()


end

function GameObjCache:__clear()

end

function GameObjCache:setup_event()

end

function GameObjCache:clear_event()


end


--//-------~★~-------~★~-------~★~GameObject加载~★~-------~★~-------~★~-------//


function GameObjCache:load_sync(url)

    local go = csGameObjCache:LoadSync(url)
    return go
end

function GameObjCache:load_async(url, on_complete, target, refer)

    if type(on_complete) ~= "function" then
        --不能没有回调
        return end

	url = string.lower(url)

	refer = refer or target
    Refer.assert(refer)    --这里纯粹为了回调

    if csAssetCache:HasAsset(url) then
        local go = csGameObjCache:CreateGo(url, nil)
        safe_call(on_complete, target, go, url)
	    --on_complete(refer, go, url)
        return
    end
    
    local evtType = url
    self:attach(evtType, on_complete, target, refer)     --自己监听加载完成事件
    
    assetCache:load_async(url, self.on_go_loaded, nil)         --注意:这里没有传引用者,因此在加载完成之前,这资源未被持有
end

--移除go加载监听
function GameObjCache:detach_load(url, target)

	url = string.lower(url)
    local evtType = url
    self:detachByType2(evtType, target)
end

--@url 已经是小写路径
function GameObjCache.on_go_loaded(prefab, url)

	--nslog.print_t('on_go_loaded', prefab, url)

    local function back(tar, fn)
        local go = csGameObjCache:CreateGo(url, nil) --创建go
        --safe_call(fn, tar, go, url)
	    fn( tar, go, url )
    end

    local evtType = url
    self.m_notifier:notifyEach(evtType, back)
    self:detachByType(evtType)  --移除所有监听
end


return GameObjCache

File: mg01/game/tech/Assets/Script/Org/Resource/Cache/GameObjCache.cs
/* ==============================================================================
 * gameobject缓存
 * @author jr.zeng
 * 2017/6/6 15:23:33
 * ==============================================================================*/

using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;


using UnityEngine;
using Object = UnityEngine.Object;


namespace mg.org
{

    public class GameObjCache : CCModule
    {
        static GameObjCache __me;
        public static GameObjCache me
        {
            get
            {
                if (__me == null)
                    __me = new GameObjCache();
                return __me;
            }
        }


        AssetCache m_assetCache = AssetCache.me;

        ClassPool2<LoadItem> m_itemPool = ClassPools.me.CreatePool<LoadItem>();
        
        Dictionary<string, List<GameObject>> m_url2goList = new Dictionary<string, List<GameObject>>();

        List<LoadItem> m_itemList = new List<LoadItem>();

        public GameObjCache()
        {
            m_notifier = new Subject();  //独立观察者
        }

        override protected void __Setup(params object[] params_)
        {
            base.__Setup();

        }

        override protected void __Clear()
        {
            m_assetCache.ReleaseByRefer(this);
            m_itemList.Clear();
            m_itemPool.Clear();
        }
        

        //-------∽-★-∽------∽-★-∽--------∽-★-∽加载操作∽-★-∽--------∽-★-∽------∽-★-∽--------//

        public GameObject LoadSync(string url_)
        {
            url_ = url_.ToLower();
            m_assetCache.LoadSync( url_, this);  //this也做引用者
            return CreateGo(url_, null);
        }

        /// <summary>
        /// 异步加载
        /// </summary>
        /// <param name="url_"></param>
        /// <param name="onComplete_"></param>
        /// <param name="refer_">只影响onComplete,不会用来持有资源</param>
        public void LoadAsync(string url_, CALLBACK_GO onComplete_, object refer_)
        {
            url_ = url_.ToLower();
            m_assetCache.RetainByUrl(this, url_);
            //m_assetCache.Retain(refer_, url_);    //引用者只用this

            if (m_assetCache.HasAsset(url_))
            {
                CreateGo(url_, onComplete_);
                return;
            }

            string referId = Refer.Format(refer_);
            AddLoadItem( url_, onComplete_, referId);

            m_assetCache.LoadAsync( url_, OnLoaded, this);  //已自己作为引用者,在ReleaseUnused时才有可能释放
        }
        

        void OnLoaded(object data_)
        {
            AssetData data = data_ as AssetData;

            if(m_itemList.Count > 0)
            {
                LoadItem item;
                for (int i = m_itemList.Count - 1; i >= 0; --i)
                {
                    item = m_itemList[i];
                    if (item.url == data.url)
                    {
                        CreateGo(item.url, item.onComplete);
                        RemoveLoadItem(item);
                    }
                }
                
            }
            
        }



        //-------∽-★-∽------∽-★-∽--------∽-★-∽gameobject管理∽-★-∽--------∽-★-∽------∽-★-∽--------//

        public GameObject CreateGo(string url_, CALLBACK_GO onComplete_)
        {
            url_ = url_.ToLower();

            GameObject prefab = m_assetCache.GetAsset(url_) as GameObject;
            if (prefab == null)
            {
                Log.Assert("gameobject未加载:" + url_);
                return null;
            }

            GameObject go = GameObjUtil.CreateGameObj(prefab);

            GameObjUtil.DontDestroyOnLoad(go);  //所有加载出来的go都是默认不销毁

            AddGo(url_, go);

            if (onComplete_ != null)
                onComplete_(go);

            return go;
        }


        void AddGo(string url_, GameObject go_)
        {
            List<GameObject> list;
            if (!m_url2goList.TryGetValue(url_, out list))
            {
                list = new List<GameObject>();
                m_url2goList[url_] = list;
            }

            list.Add(go_);
        }

        /// <summary>
        /// 释放无引用的资源
        /// </summary>
        public void ReleaseUnused()
        {
            if (m_url2goList.Count == 0)
                return;

            List<string> delList = null;
            List<GameObject> goList;

            foreach (var kvp in m_url2goList)
            {
                goList = kvp.Value;

                for (int i = goList.Count - 1; i >= 0; --i)
                {
                    if (goList[i])
                    {
                    }
                    else
                    {
                        //已被destroy
                        goList.RemoveAt(i);
                    }
                }

                if (goList.Count == 0)
                {
                    //全部销毁了, 释放引用
                    if (delList == null)
                        delList = new List<string>();
                    delList.Add(kvp.Key);

                    AssetCache.me.ReleaseByUrl(this, kvp.Key);
                }
            }

            //字典删除方案
            if (delList != null)
            {
                DicUtil.RemoveByKeys<string, List<GameObject>>(m_url2goList, delList);
            }
        }


        //-------∽-★-∽------∽-★-∽--------∽-★-∽LoadItem∽-★-∽--------∽-★-∽------∽-★-∽--------//

        LoadItem AddLoadItem(string url_, CALLBACK_GO onComplete_, string referId_)
        {
            LoadItem item = m_itemPool.Pop();
            item.Init( url_, onComplete_, referId_);

            m_itemList.Add(item);

            if(referId_ != null)
            {
                Refer.AttachDeactive(referId_, onDeactive); //监听沉默
            }

            return item;
        }

        void onDeactive(object referId_)
        {
            if (m_itemList.Count == 0)
                return;

            string referId = referId_ as string;
            for (int i = m_itemList.Count - 1; i >=0; --i)
            {
                if(m_itemList[i].referId == referId)
                {
                    RemoveLoadItem(m_itemList[i]);
                }
            }
        }

        void RemoveLoadItem(LoadItem item_)
        {
            if(item_.referId != null)
            {
                Refer.DetachDeactive(item_.referId, onDeactive);
            }

            item_.Clear();

            m_itemList.Remove(item_);
            m_itemPool.Push(item_);
        }

        class LoadItem
        {
            public string url;
            public CALLBACK_GO onComplete;
            public string referId;

            public LoadItem() { }
            
            public void Init(string url_, CALLBACK_GO onComplete_, string referId_)
            {
                url = url_;
                onComplete = onComplete_;
                referId = referId_;
            }

            public void Clear()
            {
                url = null;
                onComplete = null;
                referId = null;
            }


        }

    }




}
